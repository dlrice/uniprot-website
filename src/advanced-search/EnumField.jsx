import React from 'react';

const EnumField = ({ field, handleChange }) => (
  <div className="advanced-search__inputs" key={field.term}>
    <label htmlFor={`select_${field.term}`}>
      {field.label}
      <select
        onChange={e => handleChange(e.target.value)}
        id={`select_${field.term}`}
      >
        {field.values
          && field.values.map(item => (
            <option value={item.value} key={`select_${item.value}`}>
              {item.name}
            </option>
          ))}
      </select>
    </label>
  </div>
);

export default EnumField;
