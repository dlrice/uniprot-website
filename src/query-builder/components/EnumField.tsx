import React, { useEffect, useState } from 'react';
import { QueryBit, SearchTermType } from '../types/searchTypes';

const initializer = (field: SearchTermType, initialValue: QueryBit) => () => {
  if (!(initialValue && initialValue[field.term])) {
    console.log(field.term, 'no initial value');
    return '';
  }
  const match = new RegExp(`^\\(${'fragment'}:(\\w*)\\)$`).exec(
    initialValue[field.term]
  );
  if (!match) {
    console.log(field.term, 'invalid queryBit');
    return '';
  }
  const [, value] = match;
  if (
    field.regex &&
    field.regex !== undefined &&
    !new RegExp(field.regex).test(value)
  ) {
    console.log(field.term, 'invalid value');
    return '';
  }
  return value;
};

const EnumField: React.FC<{
  field: SearchTermType;
  handleChange: (queryBit: QueryBit) => void;
  initialValue?: QueryBit;
}> = ({ field, handleChange, initialValue }) => {
  // should initialValue be initialised to the first item?
  const [value, setValue] = useState(initializer(field, initialValue));

  useEffect(() => {
    if (value.length > 0) {
      handleChange({ [field.id]: `(${field.term}:${value?.trim()})` });
    }
  }, [field, value, handleChange]);

  return (
    <label htmlFor={`select_${field.term}`}>
      {field.label}
      <select
        onChange={(e) => setValue(e.target.value)}
        id={`select_${field.term}`}
        data-testid="enum-field-select"
        value={value}
      >
        {field.values &&
          field.values.map((item) => (
            <option value={item.value} key={`select_${item.value}`}>
              {item.name}
            </option>
          ))}
      </select>
    </label>
  );
};

export default EnumField;
