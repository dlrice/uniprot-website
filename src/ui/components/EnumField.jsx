import React from 'react';

const EnumField = ({ term, handleInputChange }) => (
  <div className="advanced-search__inputs" key={term.term}>
    <label htmlFor={`select_${term.term}`}>
      {term.label}
      <select onChange={handleInputChange} id={`select_${term.term}`}>
        {term.values
          && term.values.map(item => (
            <option value={item.value} key={`select_${item.value}`}>
              {item.name}
            </option>
          ))}
      </select>
    </label>
  </div>
);

export default EnumField;
