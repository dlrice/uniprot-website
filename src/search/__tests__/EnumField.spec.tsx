import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import EnumField from '../EnumField';

let props;
let rendered;

describe('Enum field', () => {
  beforeEach(() => {
    props = {
      field: {
        label: 'Protein Existence [PE]',
        itemType: 'single',
        term: 'existence',
        dataType: 'enum',
        values: [
          {
            name: 'Evidence at protein level',
            value: '1',
          },
          {
            name: 'Evidence at transcript level',
            value: '2',
          },
        ],
        description: 'Search by protein existence',
        example: '1',
      },
      handleChange: jest.fn(),
      queryInput: { stringValue: '1' },
    };
    rendered = render(<EnumField {...props} />);
  });

  test('should render an enum field', () => {
    const { asFragment } = rendered;
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render handle change', () => {
    const { queryAllByTestId } = rendered;
    const nodes = queryAllByTestId('enum-field-select');
    fireEvent.change(nodes[0], { target: { value: 'val1' } });
    expect(props.handleChange).toBeCalled();
  });
});
