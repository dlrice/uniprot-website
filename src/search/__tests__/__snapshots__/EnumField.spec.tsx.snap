// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enum field should render an enum field 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EnumField
    field={
      Object {
        "dataType": "enum",
        "description": "Search by protein existence",
        "example": "1",
        "itemType": "single",
        "label": "Protein Existence [PE]",
        "term": "existence",
        "values": Array [
          Object {
            "name": "Evidence at protein level",
            "value": "1",
          },
          Object {
            "name": "Evidence at transcript level",
            "value": "2",
          },
        ],
      }
    }
    handleChange={[MockFunction]}
    queryInput={
      Object {
        "stringValue": "1",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <label
        htmlFor="select_existence"
      >
        Protein Existence [PE]
        <select
          id="select_existence"
          onChange={[Function]}
        >
          <option
            value="1"
          >
            Evidence at protein level
          </option>
          <option
            value="2"
          >
            Evidence at transcript level
          </option>
        </select>
      </label>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Protein Existence [PE]",
          <select
            id="select_existence"
            onChange={[Function]}
          >
            <option
              value="1"
            >
              Evidence at protein level
            </option>
            <option
              value="2"
            >
              Evidence at transcript level
            </option>
          </select>,
        ],
        "htmlFor": "select_existence",
      },
      "ref": null,
      "rendered": Array [
        "Protein Existence [PE]",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <option
                value="1"
              >
                Evidence at protein level
              </option>,
              <option
                value="2"
              >
                Evidence at transcript level
              </option>,
            ],
            "id": "select_existence",
            "onChange": [Function],
            "value": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "select_1",
              "nodeType": "host",
              "props": Object {
                "children": "Evidence at protein level",
                "value": "1",
              },
              "ref": null,
              "rendered": "Evidence at protein level",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "select_2",
              "nodeType": "host",
              "props": Object {
                "children": "Evidence at transcript level",
                "value": "2",
              },
              "ref": null,
              "rendered": "Evidence at transcript level",
              "type": "option",
            },
          ],
          "type": "select",
        },
      ],
      "type": "label",
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <label
          htmlFor="select_existence"
        >
          Protein Existence [PE]
          <select
            id="select_existence"
            onChange={[Function]}
          >
            <option
              value="1"
            >
              Evidence at protein level
            </option>
            <option
              value="2"
            >
              Evidence at transcript level
            </option>
          </select>
        </label>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Protein Existence [PE]",
            <select
              id="select_existence"
              onChange={[Function]}
            >
              <option
                value="1"
              >
                Evidence at protein level
              </option>
              <option
                value="2"
              >
                Evidence at transcript level
              </option>
            </select>,
          ],
          "htmlFor": "select_existence",
        },
        "ref": null,
        "rendered": Array [
          "Protein Existence [PE]",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <option
                  value="1"
                >
                  Evidence at protein level
                </option>,
                <option
                  value="2"
                >
                  Evidence at transcript level
                </option>,
              ],
              "id": "select_existence",
              "onChange": [Function],
              "value": undefined,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "select_1",
                "nodeType": "host",
                "props": Object {
                  "children": "Evidence at protein level",
                  "value": "1",
                },
                "ref": null,
                "rendered": "Evidence at protein level",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": "select_2",
                "nodeType": "host",
                "props": Object {
                  "children": "Evidence at transcript level",
                  "value": "2",
                },
                "ref": null,
                "rendered": "Evidence at transcript level",
                "type": "option",
              },
            ],
            "type": "select",
          },
        ],
        "type": "label",
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
